### PR TITLE
Use frozensets for key protocols

### LINKED ISSUE #5557: Avoid defensive copying in keys protocols
Measurement and control keys protocols both do defensive copies before returning. However they are also used in hot loops where perf is important. It's likely better to specify that dunders should return return a frozenset instead, especially since most ops will return an empty one, which is a singleton requiring no allocation.

This may be impossible to do in a completely non breaking way, however I can't think of a reason users would rely on these returning mutable sets. 

Update: The mypy types of the return values is AbstractSet, not Set. And FrozenSet is also a subclass of AbstractSet so this is not breaking.

### PR DESCRIPTION
**Pull request by daxfohl**
Fixes #5557, using frozenset in key protocol dunder methods, thus allowing the top-level protocol method to avoid defensive copies. Provide a temporary warning until v0.16 for third-party dunders to do the same. This speeds up tight loops about 15% in my testing. @95-martin-orion 

Note this is not breaking, as FrozenSet is a subclass of AbstractSet. (AbstractSet is a readonly interface of a set, and is superclass of both FrozenSet and plain Set). Changed the return types to FrozenSet to be more explicit and make chaining easier.

### CODE REVIEW

File: cirq-core/cirq/protocols/measurement_key_protocol_test.py [hunk 1]
```diff
@@ -220,10 +220,10 @@ def _with_measurement_key_mapping_(self, key_map):
 def test_measurement_key_path():
     class MultiKeyGate:
         def __init__(self, keys):
-            self._keys = set([cirq.MeasurementKey.parse_serialized(key) for key in keys])
+            self._keys = frozenset([cirq.MeasurementKey.parse_serialized(key) for key in keys])
 
         def _measurement_key_names_(self):
-            return {str(key) for key in self._keys}
+            return frozenset([str(key) for key in self._keys])
```
Comment (pavoljuhas) [inline cirq-core/cirq/protocols/measurement_key_protocol_test.py:226]: please remove square brackets, no need for intermediate list.

File: cirq-core/cirq/circuits/circuit.py [hunk 2]
```diff
@@ -918,26 +918,30 @@ def qid_shape(
...
+    def all_measurement_key_objs(self) -> FrozenSet['cirq.MeasurementKey']:
+        return frozenset(
+            key for op in self.all_operations() for key in protocols.measurement_key_objs(op)
+        )
 
-    def _measurement_key_objs_(self) -> AbstractSet['cirq.MeasurementKey']:
+    def _measurement_key_objs_(self) -> FrozenSet['cirq.MeasurementKey']:
         """Returns the set of all measurement keys in this circuit.
 
         Returns: AbstractSet of `cirq.MeasurementKey` objects that are
```
Comment (pavoljuhas) [inline cirq-core/cirq/circuits/circuit.py:929]: Replace AbstractSet --> FrozenSet.  Please check return type in all other affected docstrings.

**Review by pavoljuhas — APPROVED** [review]
LGTM with fixing a few cosmetic issues.

File: cirq-core/cirq/protocols/control_key_protocol.py [hunk 3]
```diff
@@ -34,7 +35,7 @@ class SupportsControlKey(Protocol):
     """
 
     @doc_private
-    def _control_keys_(self) -> Iterable['cirq.MeasurementKey']:
+    def _control_keys_(self) -> FrozenSet['cirq.MeasurementKey']:
```
Comment (95-martin-orion) [inline cirq-core/cirq/protocols/control_key_protocol.py:38]: This is more complicated than the `AbstractSet` / `FrozenSet` case - we need to ensure that internal code doesn't take advantage of set-like behavior from this until after a release.

Comment (daxfohl) [reply cirq-core/cirq/protocols/control_key_protocol.py:38]: Yeah, this is handled by lines 60-65 in this file.

File: cirq-core/cirq/ops/gate_operation.py [hunk 4]
```diff
@@ -235,7 +235,7 @@ def _measurement_key_name_(self) -> Optional[str]:
             return getter()
         return NotImplemented
 
-    def _measurement_key_names_(self) -> Optional[AbstractSet[str]]:
+    def _measurement_key_names_(self) -> Optional[FrozenSet[str]]:
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/ops/gate_operation.py:238]: Seems to be handled by line 181 of measurement_key_protocol.py below. Looks like the full expected type is really `Union[FrozenSet[str], None, NotImplementedType]`. Should I change that in the protocol declaration? Looks like I copied and pasted the same into control_keys_protocol.

File: cirq-core/cirq/protocols/control_key_protocol.py (hunk 3 above)
Comment (daxfohl) [reply cirq-core/cirq/protocols/control_key_protocol.py:38]: Is this what you meant in your top-level comment? Old external magic will get this warning. Nothing specifically relies on this frozenness, at least not that I know of.

File: cirq-core/cirq/ops/gate_operation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (95-martin-orion) [reply cirq-core/cirq/ops/gate_operation.py:238]: Yes, I think that type would be more accurate. I'm mostly confused why mypy is accepting it...

**Review by 95-martin-orion — APPROVED** [review]
LGTM modulo protocol type fixes.

**Review by 95-martin-orion — APPROVED** [review]
LGTM

**Pull request merged**